Ataxia

ihaywood3 opened the pull request:
This is Ataxia, a proposed alternative/replacement for Staggered Orders. (happy to change the name)
See my brief "whitepaper" [ataxia.pdf](https://github.com/Codaone/DEXBot/files/2152786/ataxia.pdf)
this code is slightly different: the market last price is used for the centre price, unless this isn't available, in which case it uses the midpoint of the range, so it can do true bootstrapping (i.e form orders on empty markets)

MarkoPaasila commented:
@ihaywood3 do you think it would be ok to leave Ataxia without a GUI until we have UI autogeneration? Just cli.

mikakoi commented on dexbot/strategies/ataxia.py:
```diff
@@ -0,0 +1,277 @@
+import math
+import time
+
+from bitshares.amount import Amount
```
Unused

mikakoi commented on dexbot/strategies/ataxia.py:
```diff
@@ -0,0 +1,277 @@
+    def check_at_price(self, price):
+        """True if no order in self.orderlist at this price"""
+        for o in self.orders:
+            if abs(o['price']-price)/price < 0.001:  # "within 0.1% means equal" as slight errors creep in due to rounding
```
The line exceeds 120 characters

mikakoi commented on dexbot/strategies/ataxia.py:
```diff
@@ -0,0 +1,277 @@
+            ask = float(ticker['lowestAsk'])
+            if ask > 0.0:
+                highest_buy = min(highest_buy, ask)
+        return (highest_buy, lowest_sell)
```
Redundant parentheses

mikakoi commented on dexbot/strategies/ataxia.py:
```diff
@@ -0,0 +1,277 @@
+        (maybe external price feed, maybe some other market analysis)
+        In this implementation, just return the values unmodified
+        """
+        return (self.upper_bound, self.lower_bound)
```
Redundant parentheses

mikakoi commented on dexbot/strategies/ataxia.py:
```diff
@@ -0,0 +1,277 @@
+            for price, ladder_size in downladder:
+                if price > lowest_sell:
+                    if self.check_at_price(1/price):  # sell orders are inverted
+                        total_amount = size*ladder_size  # ladder sizes from base of 1, multiple by size toget "real" order size
```
The line exceeds 120 characters

mikakoi commented on dexbot/strategies/ataxia.py:
```diff
@@ -0,0 +1,277 @@
+            self.reassess()
+
+    @classmethod
+    def get_required_assets(cls, *args, **kwargs):
```
What's the point of this method?

mikakoi commented on dexbot/strategies/ataxia.py:
```diff
@@ -0,0 +1,277 @@
+        self.expiration = 60*60*24*365*5
+        self.last_check = 0
+
+        if kwargs.get('active', True):
```
This isn't defined anywhere?

mikakoi commented:
@ihaywood3 I made some comments but this also requires changes to the configure method. I merged the UI auto generator code to master which changes the ConfigElement structure. If you're able to make the needed changes I'll be able to merge this.

ihaywood3 commented:
In Staggered Orders a static method is used to compute required assets
during configuration
I found this got too complex for ataxia and IMHO it was better to
instantiate as an 'inactive' non-trading strategy
object, then the same routines cna be used to compute the required assets
for the given settings,
this is not finished and happy to remove it but it's an aide memoire to use
this approach when/if this is required,

On Fri, Jul 27, 2018 at 9:36 PM, Mika Koivistoinen <notifications@github.com
> wrote:

> *@mikakoi* commented on this pull request.
> ------------------------------
>
> In dexbot/strategies/ataxia.py
> <https://github.com/Codaone/DEXBot/pull/223#discussion_r205745860>:
>
> > +
> +    def __init__(self, *args, **kwargs):
> +        super().__init__(*args, **kwargs)
> +
> +        self.worker_name = kwargs.get('name')
> +        self.view = kwargs.get('view')
> +        self.size = self.worker['size']
> +        self.spread = self.worker['spread'] / 100
> +        self.increment = self.worker['increment'] / 100
> +        self.upper_bound = self.worker['upper_bound']
> +        self.lower_bound = self.worker['lower_bound']
> +        # Order expiration time, should be high enough
> +        self.expiration = 60*60*24*365*5
> +        self.last_check = 0
> +
> +        if kwargs.get('active', True):
>
> This isn't defined anywhere?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Codaone/DEXBot/pull/223#pullrequestreview-141078433>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAuTOcqwLZQaUwO58pisQj6antdxj8CMks5uKvsigaJpZM4U-Tqs>
> .
>

ihaywood3 commented:
respectfully I disagree I find readability suffers without parentheses.

OK I'm betraying my LISP/Prolog influences: I just like parentheses

On Fri, Jul 27, 2018 at 9:05 PM, Mika Koivistoinen <notifications@github.com
> wrote:

> *@mikakoi* commented on this pull request.
> ------------------------------
>
> In dexbot/strategies/ataxia.py
> <https://github.com/Codaone/DEXBot/pull/223#discussion_r205739801>:
>
> > +            centre = float(ticker['latest'])
> +        else:
> +            # there's no latest price: we are bootstrapping so use the midpoint of the range
> +            centre = (upper_bound + lower_bound)/2
> +        lowest_sell = centre * (1 + (spread/2))
> +        highest_buy = centre * (1 - (spread/2))
> +        # don't ever trade against the market
> +        if 'highestBid' in ticker and ticker['highestBid']:
> +            bid = float(ticker['highestBid'])
> +            if bid > 0.0:
> +                lowest_sell = max(lowest_sell, bid)
> +        if 'lowestAsk' in ticker and ticker['lowestAsk']:
> +            ask = float(ticker['lowestAsk'])
> +            if ask > 0.0:
> +                highest_buy = min(highest_buy, ask)
> +        return (highest_buy, lowest_sell)
>
> Redundant parentheses
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Codaone/DEXBot/pull/223#pullrequestreview-141071124>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAuTOYLTj3hKUD7vHQkSvvtgCjvgsAPqks5uKvPfgaJpZM4U-Tqs>
> .
>

ihaywood3 commented:
ACK
will let you know when I have updated the config metadata and fixed what
autopep8 missed.

On Fri, Jul 27, 2018 at 9:47 PM, Mika Koivistoinen <notifications@github.com
> wrote:

> @ihaywood3 <https://github.com/ihaywood3> I made some comments but this
> also requires changes to the configure method. I merged the UI auto
> generator code to master which changes the ConfigElement structure. If
> you're able to make the needed changes I'll be able to merge this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Codaone/DEXBot/pull/223#issuecomment-408395533>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAuTOXkowDgP0N_6GnLuPO30nAazypxgks5uKv3lgaJpZM4U-Tqs>
> .
>

mikakoi commented:
You're free to use the parentheses if you want to, I just pointed out that they're redundant.

mikakoi commented:
@ihaywood3 Did you just accidentally the whole repo?

ihaywood3 commented:
I tried to merge changes but caused more problems than I solved.
I have used git reset --hard to hopefully undo all this and just uploaded
my latest version of ataxia

On Thu, Aug 2, 2018 at 10:31 PM, Mika Koivistoinen <notifications@github.com
> wrote:

> @ihaywood3 <https://github.com/ihaywood3> Did you just accidentally the
> whole repo?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Codaone/DEXBot/pull/223#issuecomment-409909698>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAuTOfmIaoQh3eQW2fVImD940CShjRhrks5uMvELgaJpZM4U-Tqs>
> .
>

mikakoi commented:
You added bunch of changes that aren't related to ataxia, could you remove those changes?
